src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -0,0 +1,459 @@
+
+  arma::mat cij = arma::zeros(dataset.n_rows, dataset.n_rows);
+  arma::mat cil = arma::zeros(dataset.n_rows, dataset.n_rows);
+
+  for (size_t i = 0; i < dataset.n_cols; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate gradient due to target neighbors.
+      arma::vec diff = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+      cij += diff * arma::trans(diff);
+    }
+
+    for (int j = k - 1; j >= 0; j--)
+    {
+      // Bound constraints to avoid uneccesary computation.
+      for (size_t l = 0, bp = k; l < bp ; l++)
+      {
+        // Calculate gradient due to triplets.
+        double eval = metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(targetNeighbors(j, i))) -
+                      metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(impostors(l, i)));
+
+        // Check bounding condition.
+        if (eval < -1)
+        {
+          // update bound.
+          bp = l;
+          break;
+        }
+
+        arma::vec diff = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+        cil += diff * arma::trans(diff);
+
+        diff = dataset.col(i) - dataset.col(impostors(l, i));
+        cil -= diff * arma::trans(diff);
+      }
+    }
+  }
+
+  gradient = (1 - regularization) * cij + regularization * cil;
+}
+
+//! Compute gradient over a batch of data points.
+template<typename MetricType>
+template<typename GradType>
+void LMNNFunction<MetricType>::Gradient(const arma::mat& /* coordinates */,
+                                        const size_t begin,
+                                        GradType& gradient,
+                                        const size_t batchSize)
+{
+  gradient.zeros(dataset.n_rows, dataset.n_rows);
+
+  arma::mat cij = arma::zeros(dataset.n_rows, dataset.n_rows);
+  arma::mat cil = arma::zeros(dataset.n_rows, dataset.n_rows);
+
+  for (size_t i = begin; i < begin + batchSize; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate gradient due to target neighbors.
+      arma::vec diff = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+      cij += diff * arma::trans(diff);
+    }
+
+    for (int j = k - 1; j >= 0; j--)
+    {
+      // Bound constraints to avoid uneccesary computation.
+      for (size_t l = 0, bp = k; l < bp ; l++)
+      {
+        // Calculate gradient due to triplets.
+        double eval = metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(targetNeighbors(j, i))) -
+                      metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(impostors(l, i)));
+
+        // Check bounding condition.
+        if (eval < -1)
+        {
+          // update bound.
+          bp = l;
+          break;
+        }
+
+        arma::vec diff = dataset.col(i) - dataset.col(targetNeighbors(j, i));
+        cil += diff * arma::trans(diff);
```

Comment:
As a follow-up on the low-rank idea, I think this bit will need to be re-derived.  I'll take a look into that after I look into the bounds for the impostors that you wrote up.

Reply:
Now, if we partially derivate the gradient w.r.t L (instead of M) we may just get an extra factor of multiple of L (coordinates). As far I could see all other things should remain same. Though, I am not sure about this.

Reply:
I think you are right, but I haven't double-checked yet.

Reply:
What I computed for the quantity df/dL:

```
df/dL = (1 - mu) * \sum_{i, j} 2 L (x_i - x_j) (x_i - x_j)^T + \sum_{i, j, l \in N} (2 L (x_i - x_j) (x_i - x_j)^T - 2 L (x_i - x_l) (x_i - x_l)^T).
```

Full whiteboard derivation attached:
![img_20180604_160755](https://user-images.githubusercontent.com/1845039/40939057-936d701c-6811-11e8-99a7-257629192f80.jpg)

Reply:
Note that the quantity is actually a little different than before, we can't just use `transformedDataset` to compute the gradient.  Actually we have to compute the gradient in the original space and multiply `L`.  That also means we can precompute the quantities `sum_{i, j} (x_i - x_j)(x_i - x_j)^T` and probably the second term but there may be a little more trickiness there.

Reply:
Right, I think the formulation you made is absolutely correct, Earlier I also got the same exact formulation. I think we can rely on this.

Reply:
>  we can precompute the quantities sum_{i, j} (x_i - x_j)(x_i - x_j)^T and probably the second term but there may be a little more trickiness there.

We can surely precalculate the first term here but as you already said, precomputing second part is not so straightforward and I think it might not be possible even. 

I tried the above formulation and results are fine but I am facing some obstacles:
1. For a comparatively large value of step_size objective keep on increasing to inf.
2. At some value of step_size the objective decreases to minima and they again slowly starts to increase leaving final objective greater than earlier minima.
3. This shouldn't be happening but at some value of k, final accuracy is less than initial ones.

Reply:
Can you push the implementation?  One note is in the derivation above, the second term only applies to those triplets i,j,k where `1 + (x_i - x_j)^T L^T L (x_i - x_j) - (x_i - x_l)^T L^T L (x_i - x_l))` is greater than 0.

Reply:
I just pushed the changes regarding lmnn_function_impl.hpp, please have a look at them.

Reply:
I took a look and I added some numeric gradient checker tests like we do for the ANN layers, and everything seemed fine:

```
diff --git a/src/mlpack/tests/lmnn_test.cpp b/src/mlpack/tests/lmnn_test.cpp
index 43020988a..f021a0595 100644
--- a/src/mlpack/tests/lmnn_test.cpp
+++ b/src/mlpack/tests/lmnn_test.cpp
@@ -370,4 +370,113 @@ BOOST_AUTO_TEST_CASE(LMNNLBFGSSimpleDataset)
   BOOST_REQUIRE_LT(finalObj, initObj);
 }
 
+// Simple numerical gradient checker.
+template<class FunctionType>
+double CheckGradient(FunctionType& function,
+                     arma::mat& coordinates,
+                     const double eps = 1e-7)
+{
+  // Get gradients for the current parameters.
+  arma::mat orgGradient, gradient, estGradient;
+  function.Gradient(coordinates, orgGradient);
+
+  estGradient = arma::zeros(orgGradient.n_rows, orgGradient.n_cols);
+
+  // Compute numeric approximations to gradient.
+  for (size_t i = 0; i < orgGradient.n_elem; ++i)
+  {
+    double tmp = coordinates(i);
+
+    // Perturb parameter with a positive constant and get costs.
+    coordinates(i) += eps;
+    double costPlus = function.Evaluate(coordinates);
+
+    // Perturb parameter with a negative constant and get costs.
+    coordinates(i) -= (2 * eps);
+    double costMinus = function.Evaluate(coordinates);
+
+    // Restore the parameter value.
+    coordinates(i) = tmp;
+
+    // Compute numerical gradients using the costs calculated above.
+    estGradient(i) = (costPlus - costMinus) / (2 * eps);
+  }
+
+  // Estimate error of gradient.
+  return arma::norm(orgGradient - estGradient) /
+      arma::norm(orgGradient + estGradient);
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest)
+{
+  // Useful but simple dataset with six points and two classes.
+  arma::mat dataset        = "-0.1 -0.1 -0.1  0.1  0.1  0.1;"
+                             " 1.0  0.0 -1.0  1.0  0.0 -1.0 ";
+  arma::Row<size_t> labels = " 0    0    0    1    1    1   ";
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(2, 2, arma::fill::randu);
+
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest2)
+{
+  // Useful but simple dataset with six points and two classes.
+  arma::mat dataset        = "-0.1 -0.1 -0.1  0.1  0.1  0.1;"
+                             " 1.0  0.0 -1.0  1.0  0.0 -1.0 ";
+  arma::Row<size_t> labels = " 0    0    0    1    1    1   ";
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(2, 2, arma::fill::randn);
+
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest3)
+{
+  arma::mat dataset;
+  arma::Row<size_t> labels;
+  data::Load("iris.csv", dataset);
+  data::Load("iris_labels.txt", labels);
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(2, 2, arma::fill::randn);
+
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(LMNNFunctionGradientTest4)
+{
+  arma::mat dataset;
+  arma::Row<size_t> labels;
+  data::Load("iris.csv", dataset);
+  data::Load("iris_labels.txt", labels);
+
+  LMNNFunction<> lmnnfn(dataset, labels, 1, 0.6);
+
+  // 10 trials with random positions.
+  for (size_t i = 0; i < 10; ++i)
+  {
+    arma::mat coordinates(2, 2, arma::fill::randu);
+
+    CheckGradient(lmnnfn, coordinates);
+  }
+}
+
 BOOST_AUTO_TEST_SUITE_END();
```

(You could add that if you like.)

So it seems that nothing is wrong.  For the observations:

  1. This is expected, the step size for SGD has to be tuned.
  2. This may happen depending on the loss surface.
  3. This one concerns me a little bit more, but what is the dataset and parameters you are using?

I tried using the `mlpack_lmnn` program and added some bits that would compute the accuracy before and afterwards.  It is much faster when compiled without debugging symbols; the iris dataset with k=3 and SGD with default parameters takes about half a second on my machine.  But the dataset is fairly small, so usually the accuracy increase is pretty marginal (i.e., 96 -> 96.7%, etc.).

So I think it may be worth finding some larger datasets to try on, and seeing how it does there.  But it seems to me that the implementation is correct, based on the tests and experiments I did.  Maybe with other datasets it behaves differently---I am not sure.

Reply:
> This one concerns me a little bit more, but what is the dataset and parameters you are using?

There are not many configurations where we see this behavior and as per the observation by tuning the step_size in case of sgd we can get atleast the same accuracy.
Here are some examples:
1. iris -> k =2, default accuracy - 94.6667 , lbfgs accuracy - 94, sgd accuracy  - 94.6667 (this is the only case that exist in iris)
2. vc2 -> k = 4, default accuracy - 79.7101 , lbfgs accuracy - 77.7778, sgd accuracy  - 83.0918
3. vc2 -> k = 11, default accuracy - 82.1256 , lbfgs accuracy - 81.1594, sgd accuracy (after lot of tuning)  - 82.1256 
I think it may have to do with my accuracy calcuator https://pastebin.com/MFdVDcrz
Let's say we have k = 2 and we get two different points of two different classes, so it will always select the point with lower label value as the prediction.

Reply:
Ah, ok, I think you are right.  Technically the kNN rule should select the nearest neighbor label in case of a tie.

Reply:
Do you think, we should apply some weight parameter according to distance or just neighbor position here?
```
for (size_t j=0; j < k; j++)
      Map(labels(neighbors(j,i)))++;
```
example: instead of doing  Map(labels(neighbors(j,i)))++;
we can do `Map(labels(neighbors(j,i))) += 1 / j;` or something like else.

Reply:
Either is fine, both will solve the issue you mentioned.  I'm headed to bed now, I'll answer anything else in the morning.

Reply:
I think this is the exact problem, I tried Map(labels(neighbors(j,i)))+=std::exp(1/(j+1)) and the results were awesome.